velox/exec/TableWriter.cpp

```diff
@@ -134,36 +171,86 @@ RowVectorPtr TableWriter::getOutput() {
-          (TableWriteTraits::kLifeSpanContextKey, "TaskWide")
-          (TableWriteTraits::kTaskIdContextKey, connectorQueryCtx_->taskId())
-          (TableWriteTraits::kCommitStrategyContextKey, commitStrategyToString(commitStrategy_))
-          (TableWriteTraits::klastPageContextKey, true));
-  // clang-format on
-
+  // 3. Set commitcontext column.
+  const std::string commitContext = createTableCommitContext(true);
   auto commitContextVector = std::make_shared<ConstantVector<StringView>>(
       pool(),
       numOutputRows,
       false /*isNull*/,
       VARBINARY(),
-      StringView(commitContextJson));
+      StringView(commitContext));
 
   std::vector<VectorPtr> columns = {
       writtenRowsVector, fragmentsVector, commitContextVector};
 
+  // 4. Set null statistics columns.
+  if (aggregation_ != nullptr) {
+    for (int i = TableWriteTraits::kStatsChannel; i < outputType_->size();
+         i++) {
+      columns.push_back(BaseVector::createNullConstant(
+          outputType_->childAt(i), writtenRowsVector->size(), pool()));
+    }
+  }
+
   return std::make_shared<RowVector>(
       pool(), outputType_, nullptr, numOutputRows, columns);
 }
 
+std::string TableWriter::createTableCommitContext(bool lastOutput) {
+  // clang-format off
+    return folly::toJson(
+      folly::dynamic::object
+          (TableWriteTraits::kLifeSpanContextKey, "TaskWide")
+          (TableWriteTraits::kTaskIdContextKey, connectorQueryCtx_->taskId())
+          (TableWriteTraits::kCommitStrategyContextKey, commitStrategyToString(commitStrategy_))
+          (TableWriteTraits::klastPageContextKey, lastOutput));
+  // clang-format on
+}
+
 void TableWriter::updateWrittenBytes() {
   const auto writtenBytes = dataSink_->getCompletedBytes();
   auto lockedStats = stats_.wlock();
   lockedStats->physicalWrittenBytes = writtenBytes;
 }
 
+// static
+RowVectorPtr TableWriteTraits::createAggregationStatsOutput(
+    RowTypePtr outputType,
+    RowVectorPtr aggregationOutput,
+    StringView tableCommitContext,
+    velox::memory::MemoryPool* pool) {
+  // TODO: record aggregation stats output time.
+  if (aggregationOutput == nullptr || aggregationOutput->childrenSize() == 0) {
+    return nullptr;
+  }
+  const vector_size_t numOutputRows = aggregationOutput->childAt(0)->size();
+  std::vector<VectorPtr> columns;
+  for (int channel = 0; channel < outputType->size(); channel++) {
+    if (channel < TableWriteTraits::kContextChannel) {
+      // 1. Set null rows column.
+      // 2. Set null fragments column.
+      columns.push_back(BaseVector::createNullConstant(
+          outputType->childAt(channel), numOutputRows, pool));
+      continue;
+    }
+    if (channel == TableWriteTraits::kContextChannel) {
+      // 3. Set commitcontext column.
+      columns.push_back(std::make_shared<ConstantVector<StringView>>(
+          pool,
+          numOutputRows,
+          false /*isNull*/,
+          VARBINARY(),
+          std::move(tableCommitContext)));
```

Comment:
you don't need to move a stringView and let's just copy and it is a bit fragile doing so in a loop if call this more than once.